sql/core/src/test/scala/org/apache/spark/sql/DataFrameRangeSuite.scala

```diff
@@ -127,4 +133,28 @@ class DataFrameRangeSuite extends QueryTest with SharedSQLContext {
       }
     }
   }
+
+  test("Cancelling stage in a query with Range.") {
+    val listener = new SparkListener {
+      override def onTaskStart(taskStart: SparkListenerTaskStart): Unit = {
+        Thread.sleep(100)
```

Comment:
OK, so how about this:

```scala
  test("Cancelling stage in a query with Range.") {
    val listener = new SparkListener {
      override def onJobStart(jobStart: SparkListenerJobStart): Unit = {
        eventually(timeout(10.seconds)) {
          assert(DataFrameRangeSuite.stageToKill > 0)
        }
        sparkContext.cancelStage(DataFrameRangeSuite.stageToKill)
      }
    }

    sparkContext.addSparkListener(listener)
    for (codegen <- Seq(true, false)) {
      withSQLConf(SQLConf.WHOLESTAGE_CODEGEN_ENABLED.key -> codegen.toString()) {
        DataFrameRangeSuite.stageToKill = -1
        val ex = intercept[SparkException] {
          spark.range(1000000000L).map { x =>
            DataFrameRangeSuite.stageToKill = TaskContext.get().stageId()
            x
          }.toDF("id").agg(sum("id")).collect()
        }
        ex.getCause() match {
          case null =>
            assert(ex.getMessage().contains("cancelled"))
          case cause: SparkException =>
            assert(cause.getMessage().contains("cancelled"))
          case cause: Throwable =>
            fail("Expected the cause to be SparkException, got " + cause.toString() + " instead.")
        }
      }
      eventually(timeout(20.seconds)) {
        assert(sparkContext.statusTracker.getExecutorInfos.map(_.numRunningTasks()).sum == 0)
      }
    }
  }
```

Now we should be 100% sure what exactly is getting killed. I just checked that it fails without the changes from this PR.

Do I need a new issue in Apache Jira for this?